blogs/posts/2024-08-08-map-and-nest/map-and-nest.qmd

```diff
@@ -0,0 +1,127 @@
+---
+title: "Map and Nest"
+author: "Rhian Davies"
+date: "2023-08-08"
+categories: [purrr, R, tutorial]
+knitr:
+  opts_chunk:
+    fig.height: 2.5
+    echo: true
+    dev: png
+    dev.args:
+      bg: transparent
+---
+
+I want to share a framework that I like using occasionally for data analysis. It's the nest-and-map and it's helped me countless times when I'm working with related datasets. By combining [{purrr}](https://purrr.tidyverse.org/) mapping with [{tidyr}](https://tidyr.tidyverse.org/) nesting, I can keep raw data to hand while performing over different analyses. This allows me to easily track back from a summary or plot to the original raw data, ensuring everything is linked.
+
+The main funtions we'll need are 
+
+- `tidyr::nest()`
+- `dplyr::mutate()`
+- `purrr::map()`
+- `purrr::walk()`
+
+
+## Example on NHS workforce statistics 
+
+```{r}
+#| echo: false
+source("helper-functions.R")
+```
+
+The [NHS workforce statistics](https://digital.nhs.uk/data-and-information/publications/statistical/nhs-workforce-statistics) are official statistics published monthly for England.
+
+```{r}
+staff_group <- readRDS(file = "workforce_staff_group.rds")
+```
+
+I want to perform an analysis for each of the 42 integrated care systems (ICS). The {tidyr} `nest()` function creates a list-column, where each cell contains a mini dataframe for each grouping.
+
+Let's group by ICS, and call the nested data column `raw_data`.
+```{r}
+group_by_ics <- staff_group |>
+    tidyr::nest(raw_data = -ics_name)
+```
+
+The new column is a list-column, with each cell containing an entire tibble of data relating to that individual ICS.
+
+```{r}
+#' echo: false
+head(group_by_ics)
+```
+
+We can grab these mini datasets in the usual way and explore them interactively.
+
+```{r}
+group_by_ics$raw_data[[1]]
+```
+
+Next, let's apply some simple processing, say converting absolute numbers into percentages, to each of the ICSs in turn. We use `mutate()` to create a new list-column `staff_percent` and `map()` to apply the processing function to each cell in turn. ^[In this example, we actually didn't need to nest first. We could have performed the `mutate()` step on the full dataset.]
+
+```{r}
+processed_staff <-
+group_by_ics |>
+    dplyr::mutate(
+        staff_percent = purrr::map(raw_data, convert_percent)
```

Comment:
To consider: maybe a footnote on how the function can be provided as the raw function name (like in this example) vs `~` (like in the code block at line 85) vs `\()` and `function()` (block at line 117).

Reply:
I've dropped the ~ and linked to the help page

(I only still do ~ out of habit. Can't teach an old dog new tricks)